File: Final Round/Services/JobCache.swift
import Foundation

class JobCache {
    static let shared = JobCache()
    
    private var cache: [UUID: CachedJobData] = [:]
    private var companyInfoCache: [String: CachedCompanyInfo] = [:]
    private let cacheExpirationInterval: TimeInterval = 86400 // 24 hours for jobs
    private let companyInfoExpirationInterval: TimeInterval = 604800 // 7 days for company info
    private let cacheFileURL: URL
    private let companyInfoCacheFileURL: URL
    
    private init() {
        // Set up cache file location in Documents directory
        let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        cacheFileURL = documentsPath.appendingPathComponent("job_cache.json")
        companyInfoCacheFileURL = documentsPath.appendingPathComponent("company_info_cache.json")
        
        // Load cache from disk on initialization
        loadCacheFromDisk()
        loadCompanyInfoCacheFromDisk()
    }
    
    func cacheJobs(_ jobs: [JobPost], for userId: UUID) {
        cache[userId] = CachedJobData(jobs: jobs, timestamp: Date())
        saveCacheToDisk()
    }
    
    func getCachedJobs(for userId: UUID) -> [JobPost]? {
        guard let cachedData = cache[userId] else { return nil }
        
        // Check if cache has expired
        let timeElapsed = Date().timeIntervalSince(cachedData.timestamp)
        if timeElapsed > cacheExpirationInterval {
            cache.removeValue(forKey: userId)
            saveCacheToDisk()
            return nil
        }
        
        return cachedData.jobs
    }
    
    func cacheJobsWithCategories(_ result: JobSearchResult, for userId: UUID) {
        cache[userId] = CachedJobData(
            jobs: result.jobs,
            timestamp: Date(),
            categories: result.categories
        )
        saveCacheToDisk()
    }
    
    func getCachedJobsWithCategories(for userId: UUID) -> JobSearchResult? {
        guard let cachedData = cache[userId] else { return nil }
        
        // Check if cache has expired
        let timeElapsed = Date().timeIntervalSince(cachedData.timestamp)
        if timeElapsed > cacheExpirationInterval {
            cache.removeValue(forKey: userId)
            saveCacheToDisk()
            return nil
        }
        
        guard let categories = cachedData.categories else { return nil }
        
        return JobSearchResult(categories: categories, jobs: cachedData.jobs)
    }
    
    func clearCache() {
        cache.removeAll()
        saveCacheToDisk()
    }
    
    func clearCache(for userId: UUID) {
        cache.removeValue(forKey: userId)
        saveCacheToDisk()
        print("✅ Cleared cache for user: \(userId)")
    }
    
    // MARK: - Company Info Caching
    
    func cacheCompanyInfo(_ info: String, for companyName: String) {
        let normalizedKey = companyName.lowercased().trimmingCharacters(in: .whitespaces)
        companyInfoCache[normalizedKey] = CachedCompanyInfo(info: info, timestamp: Date())
        saveCompanyInfoCacheToDisk()
        print("✅ Cached company info for: \(companyName)")
    }
    
    func getCachedCompanyInfo(for companyName: String) -> String? {
        let normalizedKey = companyName.lowercased().trimmingCharacters(in: .whitespaces)
        guard let cachedData = companyInfoCache[normalizedKey] else { return nil }
        
        // Check if cache has expired (7 days)
        let timeElapsed = Date().timeIntervalSince(cachedData.timestamp)
        if timeElapsed > companyInfoExpirationInterval {
            companyInfoCache.removeValue(forKey: normalizedKey)
            saveCompanyInfoCacheToDisk()
            print("ℹ️ Company info cache expired for: \(companyName)")
            return nil
        }
        
        print("✅ Using cached company info for: \(companyName)")
        return cachedData.info
    }
    
    func clearCompanyInfoCache() {
        companyInfoCache.removeAll()
        saveCompanyInfoCacheToDisk()
        print("✅ Cleared all company info cache")
    }
    
    // MARK: - Disk Persistence
    
    private func saveCacheToDisk() {
        do {
            let encoder = JSONEncoder()
            encoder.dateEncodingStrategy = .iso8601
            
            // Convert cache to a serializable format
            let serializableCache = cache.mapValues { data in
                SerializableCachedJobData(
                    jobs: data.jobs,
                    timestamp: data.timestamp,
                    categories: data.categories
                )
            }
            
            // Convert UUID keys to strings for JSON serialization
            let stringKeyedCache = Dictionary(uniqueKeysWithValues: 
                serializableCache.map { (key, value) in (key.uuidString, value) }
            )
            
            let data = try encoder.encode(stringKeyedCache)
            try data.write(to: cacheFileURL, options: .atomic)
            print("✅ Job cache saved to disk")
        } catch {
            print("❌ Failed to save cache to disk: \(error)")
        }
    }
    
    private func loadCacheFromDisk() {
        guard FileManager.default.fileExists(atPath: cacheFileURL.path) else {
            print("ℹ️ No cache file found on disk")
            return
        }
        
        do {
            let data = try Data(contentsOf: cacheFileURL)
            let decoder = JSONDecoder()
            decoder.dateDecodingStrategy = .iso8601
            
            let stringKeyedCache = try decoder.decode([String: SerializableCachedJobData].self, from: data)
            
            // Convert string keys back to UUIDs
            cache = Dictionary(uniqueKeysWithValues: 
                stringKeyedCache.compactMap { (key, value) -> (UUID, CachedJobData)? in
                    guard let uuid = UUID(uuidString: key) else { return nil }
                    return (uuid, CachedJobData(
                        jobs: value.jobs,
                        timestamp: value.timestamp,
                        categories: value.categories
                    ))
                }
            )
            
            print("✅ Job cache loaded from disk (\(cache.count) users)")
        } catch {
            print("❌ Failed to load cache from disk: \(error)")
        }
    }
    
    // MARK: - Company Info Disk Persistence
    
    private func saveCompanyInfoCacheToDisk() {
        do {
            let encoder = JSONEncoder()
            encoder.dateEncodingStrategy = .iso8601
            
            let serializableCache = companyInfoCache.mapValues { data in
                SerializableCachedCompanyInfo(info: data.info, timestamp: data.timestamp)
            }
            
            let data = try encoder.encode(serializableCache)
            try data.write(to: companyInfoCacheFileURL, options: .atomic)
            print("✅ Company info cache saved to disk")
        } catch {
            print("❌ Failed to save company info cache to disk: \(error)")
        }
    }
    
    private func loadCompanyInfoCacheFromDisk() {
        guard FileManager.default.fileExists(atPath: companyInfoCacheFileURL.path) else {
            print("ℹ️ No company info cache file found on disk")
            return
        }
        
        do {
            let data = try Data(contentsOf: companyInfoCacheFileURL)
            let decoder = JSONDecoder()
            decoder.dateDecodingStrategy = .iso8601
            
            let loadedCache = try decoder.decode([String: SerializableCachedCompanyInfo].self, from: data)
            
            companyInfoCache = loadedCache.mapValues { value in
                CachedCompanyInfo(info: value.info, timestamp: value.timestamp)
            }
            
            print("✅ Company info cache loaded from disk (\(companyInfoCache.count) companies)")
        } catch {
            print("❌ Failed to load company info cache from disk: \(error)")
        }
    }
}

// MARK: - Company Info Cache Structs

private struct CachedCompanyInfo {
    let info: String
    let timestamp: Date
}

private struct SerializableCachedCompanyInfo: Codable {
    let info: String
    let timestamp: Date
}

// MARK: - Job Cache Structs

private struct CachedJobData {
    let jobs: [JobPost]
    let timestamp: Date
    let categories: [String]?
    
    init(jobs: [JobPost], timestamp: Date, categories: [String]? = nil) {
        self.jobs = jobs
        self.timestamp = timestamp
        self.categories = categories
    }
}

// Codable version for disk persistence
private struct SerializableCachedJobData: Codable {
    let jobs: [JobPost]
    let timestamp: Date
    let categories: [String]?
}

File: FinalRoundFlutter/lib/services/job_cache.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:path_provider/path_provider.dart';
import '../models/job_post.dart';

/// Service for caching jobs and company info locally
/// Matches iOS JobCache.swift implementation
class JobCache {
  static final JobCache instance = JobCache._();
  JobCache._();

  // In-memory cache
  final Map<String, _CachedJobData> _jobCache = {};
  final Map<String, _CachedCompanyInfo> _companyInfoCache = {};

  // Cache expiration times
  static const Duration jobCacheExpiration = Duration(hours: 24);
  static const Duration companyInfoExpiration = Duration(days: 7);

  // File paths
  String? _cacheDirectory;

  Future<String> get _cacheDir async {
    if (_cacheDirectory != null) return _cacheDirectory!;
    final dir = await getApplicationDocumentsDirectory();
    _cacheDirectory = dir.path;
    return _cacheDirectory!;
  }

  // MARK: - Job Caching

  /// Cache jobs for a user
  Future<void> cacheJobs(List<JobPost> jobs, String userId) async {
    _jobCache[userId] = _CachedJobData(
      jobs: jobs,
      timestamp: DateTime.now(),
    );
    await _saveJobCacheToDisk();
    debugPrint('✅ Cached ${jobs.length} jobs for user: $userId');
  }

  /// Get cached jobs for a user (returns null if expired or not found)
  List<JobPost>? getCachedJobs(String userId) {
    final cached = _jobCache[userId];
    if (cached == null) return null;

    // Check if cache has expired
    final elapsed = DateTime.now().difference(cached.timestamp);
    if (elapsed > jobCacheExpiration) {
      _jobCache.remove(userId);
      _saveJobCacheToDisk();
      debugPrint('ℹ️ Job cache expired for user: $userId');
      return null;
    }

    debugPrint('✅ Using cached jobs for user: $userId');
    return cached.jobs;
  }

  /// Cache jobs with categories
  Future<void> cacheJobsWithCategories(JobSearchResult result, String userId) async {
    _jobCache[userId] = _CachedJobData(
      jobs: result.jobs,
      timestamp: DateTime.now(),
      categories: result.categories,
    );
    await _saveJobCacheToDisk();
    debugPrint('✅ Cached ${result.jobs.length} jobs with ${result.categories.length} categories');
  }

  /// Get cached jobs with categories
  JobSearchResult? getCachedJobsWithCategories(String userId) {
    final cached = _jobCache[userId];
    if (cached == null || cached.categories == null) return null;

    final elapsed = DateTime.now().difference(cached.timestamp);
    if (elapsed > jobCacheExpiration) {
      _jobCache.remove(userId);
      _saveJobCacheToDisk();
      return null;
    }

    return JobSearchResult(
      categories: cached.categories!,
      jobs: cached.jobs,
    );
  }

  /// Clear all job cache
  void clearCache() {
    _jobCache.clear();
    _saveJobCacheToDisk();
    debugPrint('✅ Cleared all job cache');
  }

  /// Clear cache for a specific user
  void clearCacheForUser(String userId) {
    _jobCache.remove(userId);
    _saveJobCacheToDisk();
    debugPrint('✅ Cleared job cache for user: $userId');
  }

  // MARK: - Company Info Caching

  /// Cache company info
  Future<void> cacheCompanyInfo(String info, String companyName) async {
    final key = companyName.toLowerCase().trim();
    _companyInfoCache[key] = _CachedCompanyInfo(
      info: info,
      timestamp: DateTime.now(),
    );
    await _saveCompanyInfoCacheToDisk();
    debugPrint('✅ Cached company info for: $companyName');
  }

  /// Get cached company info
  String? getCachedCompanyInfo(String companyName) {
    final key = companyName.toLowerCase().trim();
    final cached = _companyInfoCache[key];
    if (cached == null) return null;

    final elapsed = DateTime.now().difference(cached.timestamp);
    if (elapsed > companyInfoExpiration) {
      _companyInfoCache.remove(key);
      _saveCompanyInfoCacheToDisk();
      debugPrint('ℹ️ Company info cache expired for: $companyName');
      return null;
    }

    debugPrint('✅ Using cached company info for: $companyName');
    return cached.info;
  }

  /// Clear company info cache
  void clearCompanyInfoCache() {
    _companyInfoCache.clear();
    _saveCompanyInfoCacheToDisk();
    debugPrint('✅ Cleared all company info cache');
  }

  // MARK: - Disk Persistence

  Future<void> _saveJobCacheToDisk() async {
    try {
      final dir = await _cacheDir;
      final file = File('$dir/job_cache.json');
      
      final data = <String, dynamic>{};
      _jobCache.forEach((key, value) {
        data[key] = {
          'jobs': value.jobs.map((j) => j.toJson()).toList(),
          'timestamp': value.timestamp.toIso8601String(),
          'categories': value.categories,
        };
      });
      
      await file.writeAsString(jsonEncode(data));
    } catch (e) {
      debugPrint('❌ Failed to save job cache: $e');
    }
  }

  Future<void> loadCacheFromDisk() async {
    try {
      final dir = await _cacheDir;
      final jobFile = File('$dir/job_cache.json');
      final companyFile = File('$dir/company_info_cache.json');

      // Load job cache
      if (await jobFile.exists()) {
        final content = await jobFile.readAsString();
        final data = jsonDecode(content) as Map<String, dynamic>;
        
        data.forEach((key, value) {
          final jobData = value as Map<String, dynamic>;
          final jobs = (jobData['jobs'] as List)
              .map((j) => JobPost.fromJson(j as Map<String, dynamic>))
              .toList();
          final timestamp = DateTime.parse(jobData['timestamp'] as String);
          final categories = (jobData['categories'] as List?)?.cast<String>();
          
          _jobCache[key] = _CachedJobData(
            jobs: jobs,
            timestamp: timestamp,
            categories: categories,
          );
        });
        debugPrint('✅ Loaded job cache from disk (${_jobCache.length} users)');
      }

      // Load company info cache
      if (await companyFile.exists()) {
        final content = await companyFile.readAsString();
        final data = jsonDecode(content) as Map<String, dynamic>;
        
        data.forEach((key, value) {
          final infoData = value as Map<String, dynamic>;
          _companyInfoCache[key] = _CachedCompanyInfo(
            info: infoData['info'] as String,
            timestamp: DateTime.parse(infoData['timestamp'] as String),
          );
        });
        debugPrint('✅ Loaded company info cache from disk (${_companyInfoCache.length} companies)');
      }
    } catch (e) {
      debugPrint('❌ Failed to load cache from disk: $e');
    }
  }

  Future<void> _saveCompanyInfoCacheToDisk() async {
    try {
      final dir = await _cacheDir;
      final file = File('$dir/company_info_cache.json');
      
      final data = <String, dynamic>{};
      _companyInfoCache.forEach((key, value) {
        data[key] = {
          'info': value.info,
          'timestamp': value.timestamp.toIso8601String(),
        };
      });
      
      await file.writeAsString(jsonEncode(data));
    } catch (e) {
      debugPrint('❌ Failed to save company info cache: $e');
    }
  }
}

/// Internal class for cached job data
class _CachedJobData {
  final List<JobPost> jobs;
  final DateTime timestamp;
  final List<String>? categories;

  _CachedJobData({
    required this.jobs,
    required this.timestamp,
    this.categories,
  });
}

/// Internal class for cached company info
class _CachedCompanyInfo {
  final String info;
  final DateTime timestamp;

  _CachedCompanyInfo({
    required this.info,
    required this.timestamp,
  });
}
